src/Nevermind/Nevermind.JsonRpc/JsonRpcService.cs

```diff
@@ -0,0 +1,207 @@
+                return GetErrorResponse(ErrorType.ParseError, "Incorrect message", null);
+            }         
+        }
+
+        private string ExecuteRequest(JsonRpcRequest rpcRequest)
+        {
+            var methodName = rpcRequest.Method.Trim().ToLower();
+            
+            var module = _moduleProvider.GetEnabledModules().FirstOrDefault(x => x.MethodDictionary.ContainsKey(methodName));
+            if (module != null)
+            {
+                return Execute(methodName, rpcRequest, module.MethodDictionary, module.ModuleObject);
+            }
+
+            return GetErrorResponse(ErrorType.MethodNotFound, $"Method {rpcRequest.Method} is not supported", rpcRequest.Id);
+        }
+
+        private string Execute(string methodName, JsonRpcRequest request, IDictionary<string, MethodInfo> methodDict, object module)
+        {
+            var method = methodDict[methodName];
+            var expectedParameters = method.GetParameters();
+            var providedParameters = request.Params;
+            if (expectedParameters.Length != (providedParameters?.Length ?? 0))
+            {
+                return GetErrorResponse(ErrorType.InvalidParams, $"Incorrect parameters count, expected: {expectedParameters.Length}", request.Id);
+            }
+
+            //prepare parameters
+            object[] parameters = null;
+            if (expectedParameters.Length > 0)
+            {
+                parameters = GetParameters(expectedParameters, providedParameters);
+                if (parameters == null)
+                {
+                    return GetErrorResponse(ErrorType.InvalidParams, "Incorrect parameters", request.Id);
+                }
+            }
+
+            //execute method
+            var result = method.Invoke(module, parameters);
+            var resultWrapper = result as IResultWrapper;
+            if (resultWrapper == null)
+            {
+                _logger.Error($"Method {methodName} execution result does not implement IResultWrapper");
+                return GetErrorResponse(ErrorType.InternalError, "Internal error", request.Id);
+            }
+            if (resultWrapper.GetResult() == null || resultWrapper.GetResult().ResultType == ResultType.Failure)
+            {
+                _logger.Error($"Error during method: {methodName} execution: {resultWrapper.GetResult()?.Error ?? "no result"}");
+                return GetErrorResponse(ErrorType.InternalError, "Internal error", request.Id);
+            }
+
+            //process response
+            var data = resultWrapper.GetData();
+            var collection = data as IEnumerable;
+            if (collection == null || data is string)
+            {
+                var json = GetDataObject(data);
+                return GetSuccessResponse(json, request.Id);        
+            }
+            var items = new List<object>();
+            foreach (var item in collection)
```

Comment:
https://stackoverflow.com/questions/365615/in-net-which-loop-runs-faster-for-or-foreach

let us do that everywhere because in this project you cannot really consider optimizations premature :)

Reply:
probably not here though since it is IEnumerable - but would like to check the root, if it can be an array